File: python/year_2015/day_24.py
"""
https://adventofcode.com/2015/day/24
"""


from pathlib import Path
from math import prod

from iteration_utilities import powerset

def partition(text, k):
    target = sum(nums := tuple(map(int, text.split()))) // k

    return prod(next(group for group in powerset(nums) if target == sum(group)))


def part_1(text, k=3):
    return partition(text, k)


def part_2(text, k=4):
    return partition(text, k)


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_24.txt").read_text()


EXAMPLE = "1 2 3 4 5 7 8 9 10 11"


def test_part_1_input():
    assert part_1(_read_input()) == 10439961859

def test_part_1_example():
    assert part_1(EXAMPLE) == 99

def test_part_2_input():
    assert part_2(_read_input()) == 72050269

def test_part_2_example_a():
    assert part_2(EXAMPLE) == 44

File: rust/year_2015/src/day_24.rs
use itertools::Itertools;

pub fn part_1(input: &str) -> u64 {
    partition(input, 3)
}

pub fn part_2(input: &str) -> u64 {
    partition(input, 4)
}

fn partition(input: &str, k: u64) -> u64 {
    let weights = parse_weights(input);
    let target = weights.iter().sum::<u64>() / k;
    weights
        .into_iter()
        .powerset()
        .find(|combination| combination.iter().sum::<u64>() == target)
        .unwrap()
        .into_iter()
        .product()
}

fn parse_weights(input: &str) -> Vec<u64> {
    input.lines().map(|line| line.parse().unwrap()).collect()
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_24.txt");

    const EXAMPLE: &str = "1
2
3
4
5
7
8
9
10
11";

    #[test_case(EXAMPLE => 99)]
    #[test_case(INPUT => 10439961859)]
    fn part_1(input: &str) -> u64 {
        super::part_1(input)
    }

    #[test_case(EXAMPLE => 44)]
    #[test_case(INPUT => 72050269)]
    fn part_2(input: &str) -> u64 {
        super::part_2(input)
    }
}
